File: packages/csharp/src/Anip.GraphQL/GraphQLAuthBridge.cs
using Anip.Core;
using Anip.Service;

namespace Anip.GraphQL;

/// <summary>
/// Per-field auth resolution for the GraphQL interface.
/// JWT-first, API-key-fallback. Subject "adapter:anip-graphql".
/// Auth errors in result body, not HTTP errors.
/// </summary>
public static class GraphQLAuthBridge
{
    /// <summary>
    /// Resolves auth from a bearer token for a GraphQL field.
    /// Same pattern as REST auth bridge but with graphql subject and source.
    /// </summary>
    public static DelegationToken ResolveAuth(string bearer, AnipService service,
                                               string capabilityName)
    {
        // Try as JWT first.
        AnipError jwtError;
        try
        {
            return service.ResolveBearerToken(bearer);
        }
        catch (AnipError e)
        {
            jwtError = e;
        }
        // Only AnipError is caught; any other exception propagates.

        // Try as API key.
        var principal = service.AuthenticateBearer(bearer);
        if (!string.IsNullOrEmpty(principal))
        {
            var capDecl = service.GetCapabilityDeclaration(capabilityName);
            List<string>? minScope = null;
            if (capDecl != null)
            {
                minScope = capDecl.MinimumScope;
            }
            if (minScope == null || minScope.Count == 0)
            {
                minScope = new List<string> { "*" };
            }

            var req = new TokenRequest
            {
                Subject = "adapter:anip-graphql",
                Scope = minScope,
                Capability = capabilityName,
                PurposeParameters = new Dictionary<string, object> { ["source"] = "graphql" },
            };

            TokenResponse tokenResp;
            try
            {
                tokenResp = service.IssueToken(principal, req);
            }
            catch
            {
                // Return the real issuance error, not the stale JWT error.
                throw;
            }

            return service.ResolveBearerToken(tokenResp.Token);
        }

        // Neither JWT nor API key -- surface the original JWT error.
        throw jwtError;
    }
}

File: packages/java/anip-graphql/src/main/java/dev/anip/graphql/GraphQLAuthBridge.java
package dev.anip.graphql;

import dev.anip.core.ANIPError;
import dev.anip.core.CapabilityDeclaration;
import dev.anip.core.DelegationToken;
import dev.anip.core.TokenRequest;
import dev.anip.core.TokenResponse;
import dev.anip.service.ANIPService;

import java.util.List;
import java.util.Map;
import java.util.Optional;

/**
 * Per-field auth resolution for the GraphQL interface.
 * JWT-first, API-key-fallback. Subject "adapter:anip-graphql".
 * Auth errors in result body, not HTTP errors.
 */
public class GraphQLAuthBridge {

    private GraphQLAuthBridge() {}

    /**
     * Resolves auth from a bearer token for a GraphQL field.
     * Same pattern as REST auth bridge but with graphql subject and source.
     */
    public static DelegationToken resolveAuth(String bearer, ANIPService service,
                                               String capabilityName) throws Exception {
        // Try as JWT first.
        ANIPError jwtError;
        try {
            return service.resolveBearerToken(bearer);
        } catch (ANIPError e) {
            jwtError = e;
        }

        // Try as API key.
        Optional<String> principal = service.authenticateBearer(bearer);
        if (principal.isPresent() && !principal.get().isEmpty()) {
            CapabilityDeclaration capDecl = service.getCapabilityDeclaration(capabilityName);
            List<String> minScope = null;
            if (capDecl != null) {
                minScope = capDecl.getMinimumScope();
            }
            if (minScope == null || minScope.isEmpty()) {
                minScope = List.of("*");
            }

            TokenRequest req = new TokenRequest(
                    "adapter:anip-graphql", minScope, capabilityName,
                    Map.of("source", "graphql"), null, 0, null
            );

            TokenResponse tokenResp;
            try {
                tokenResp = service.issueToken(principal.get(), req);
            } catch (Exception issueErr) {
                throw issueErr;
            }

            return service.resolveBearerToken(tokenResp.getToken());
        }

        throw jwtError;
    }
}
